### ISSUE TITLE
Black Flickering Screen - NullReferenceException

### ISSUE THREAD

**Issue by JordiDeMaeyer** [Dubwise56/Dubs-Central-Heating#2]
Hello,

Amidst while playing, I encountered an error.
My screen started flickering every second or so.
I looked in the log and saw the following errorlog:

`Root level exception in Update(): System.NullReferenceException: Object reference not set to an instance of an object
  at DubsCentralHeating.HygienePipeMapComp+<>c__DisplayClass30_2.<RefreshInternetsOnTile>b__4 (Verse.IntVec3 cell) [0x00000] in <b8359cc4833c48838c1e239a6bcf9a45>:0 
  at System.Linq.Enumerable.Any[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate) [0x0002c] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at DubsCentralHeating.HygienePipeMapComp.RefreshInternetsOnTile (System.Int32 tile) [0x0010d] in <b8359cc4833c48838c1e239a6bcf9a45>:0 
  at DubsCentralHeating.HygienePipeMapComp.RegenGrids () [0x00036] in <b8359cc4833c48838c1e239a6bcf9a45>:0 
  at DubsCentralHeating.Patches.HarmonyPatches_Plumbing.GridRegen_Prefix (Verse.Map __instance) [0x00006] in <b8359cc4833c48838c1e239a6bcf9a45>:0 
  at (wrapper dynamic-method) Verse.Map.Verse.Map.MapUpdate_Patch2(Verse.Map)
  at Verse.Game.UpdatePlay () [0x00031] in <81af9f8a18324e17b286924a43555a7c>:0 
  at Verse.Root_Play.Update () [0x00032] in <81af9f8a18324e17b286924a43555a7c>:0 
UnityEngine.StackTraceUtility:ExtractStackTrace ()
Verse.Log:Error (string)
Verse.Root_Play:Update ()
`

I also noticed that my radiators were not getting heating units.

If i flick power off on the boilers:

`JobDriver threw exception in initAction for pawn Brabe driver=JobDriver_Flick (toilIndex=2) driver.job=(Flick (Job_9431744) A=Thing_ElectricBoiler3470656)
System.NullReferenceException: Object reference not set to an instance of an object
  at DubsCentralHeating.HygienePipeMapComp+<>c__DisplayClass30_2.<RefreshInternetsOnTile>b__4 (Verse.IntVec3 cell) [0x00000] in <b8359cc4833c48838c1e239a6bcf9a45>:0 
  at System.Linq.Enumerable.Any[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate) [0x0002c] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at DubsCentralHeating.HygienePipeMapComp.RefreshInternetsOnTile (System.Int32 tile) [0x0010d] in <b8359cc4833c48838c1e239a6bcf9a45>:0 
  at DubsCentralHeating.HygienePipeMapComp.RegenGrids () [0x00036] in <b8359cc4833c48838c1e239a6bcf9a45>:0 
  at DubsCentralHeating.CompPipe.ReceiveCompSignal (System.String signal) [0x00042] in <b8359cc4833c48838c1e239a6bcf9a45>:0 
  at Verse.ThingWithComps.BroadcastCompSignal (System.String signal) [0x0002b] in <81af9f8a18324e17b286924a43555a7c>:0 
  at RimWorld.CompFlickable.set_SwitchIsOn (System.Boolean value) [0x0002b] in <81af9f8a18324e17b286924a43555a7c>:0 
  at RimWorld.CompFlickable.DoFlick () [0x00007] in <81af9f8a18324e17b286924a43555a7c>:0 
  at RimWorld.JobDriver_Flick+<>c__DisplayClass1_0.<MakeNewToils>b__1 () [0x00046] in <81af9f8a18324e17b286924a43555a7c>:0 
  at Verse.AI.JobDriver.TryActuallyStartNextToil () [0x001de] in <81af9f8a18324e17b286924a43555a7c>:0 
UnityEngine.StackTraceUtility:ExtractStackTrace ()
Verse.Log:Error (string)
Verse.AI.JobUtility:TryStartErrorRecoverJob (Verse.Pawn,string,System.Exception,Verse.AI.JobDriver)
Verse.AI.JobDriver:TryActuallyStartNextToil ()
(wrapper dynamic-method) Verse.AI.JobDriver:Verse.AI.JobDriver.ReadyForNextToil_Patch1 (Verse.AI.JobDriver)
(wrapper dynamic-method) Verse.AI.JobDriver:Verse.AI.JobDriver.DriverTick_Patch0 (Verse.AI.JobDriver)
Verse.AI.Pawn_JobTracker:JobTrackerTick ()
(wrapper dynamic-method) Verse.Pawn:Verse.Pawn.Tick_Patch2 (Verse.Pawn)
Verse.TickList:Tick ()
(wrapper dynamic-method) Verse.TickManager:Verse.TickManager.DoSingleTick_Patch3 (Verse.TickManager)
Verse.TickManager:TickManagerUpdate ()
Verse.Game:UpdatePlay ()
Verse.Root_Play:Update ()
`

On this pastebin, you can find the error on a 'clean' load of the game.
[https://pastebin.com/HUmR4sc1](Player.log)

After some experimenting: just the boilers Connected/capacity 0/2250, slowly adding ratiator after radiator, at some point on a newly constructed radiator, all the heating units drop and again its connected/capacity is at 0/0 again.

uninstalled central heating, installed bad hygene and reconnected everything, and it works again.
went back to central heating and same problem occurs.

uninstalled central heating, loaded the game, saved, installed central heating and on the placement of the first boiler I get the first errormessage agian.

Any idea what might be the problem/fix?

haven't installed new mods in a while, not sure if any updated recently

**Comment by Dubwise56** [comment]
hmm do you have a mod that adds multiple maps per tile, like extra floors, mines, encounter maps, space maps, anything like this

**Comment by JordiDeMaeyer** [comment]
That appears to do the trick.

I dissabled DeepRim and tested on a new save. No problems anymore.
Is there any work around that you know of? 
Or just: "Too bad, no compatibiliyt"?

Either way, Thank you for the help and reply!

**Comment by Dubwise56** [comment]
well i added a type of universal support for any mod that creates new levels, so it should work unless there's a malformed floor or something

**Comment by JordiDeMaeyer** [comment]
Oke, Abandoning the Deep Rim Layer fixed this problem!
there was a heat pipe grid on that layer aswel, a copy of the ground layer.

Thank you for the help! :)

**Comment by estafeit** [comment]
Hello,

I have the same beahavior with black screen flickering when working on heating/cooling grid, assorted with the very same NullReferenceExceptions. I'm also using DeepRim.

`Root level exception in Update(): System.NullReferenceException: Object reference not set to an instance of an object
  at DubsCentralHeating.HygienePipeMapComp+<>c__DisplayClass30_2.<RefreshInternetsOnTile>b__4 (Verse.IntVec3 cell) [0x00000] in <b8359cc4833c48838c1e239a6bcf9a45>:0 
  at System.Linq.Enumerable.Any[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate) [0x0002c] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at DubsCentralHeating.HygienePipeMapComp.RefreshInternetsOnTile (System.Int32 tile) [0x0010d] in <b8359cc4833c48838c1e239a6bcf9a45>:0 
  at DubsCentralHeating.HygienePipeMapComp.RegenGrids () [0x00036] in <b8359cc4833c48838c1e239a6bcf9a45>:0 
  at DubsCentralHeating.Patches.HarmonyPatches_Plumbing.GridRegen_Prefix (Verse.Map __instance) [0x00006] in <b8359cc4833c48838c1e239a6bcf9a45>:0 
  at (wrapper dynamic-method) Verse.Map.Verse.Map.MapUpdate_Patch3(Verse.Map)
  at (wrapper dynamic-method) Verse.Game.Verse.Game.UpdatePlay_Patch2(Verse.Game)
  at (wrapper dynamic-method) Verse.Root_Play.Verse.Root_Play.Update_Patch1(Verse.Root_Play)
UnityEngine.StackTraceUtility:ExtractStackTrace ()
Verse.Log:Error (string)
(wrapper dynamic-method) Verse.Root_Play:Verse.Root_Play.Update_Patch1 (Verse.Root_Play)
`

Once I tried installing an air-con on deeprim layer, real problems started, with my main network malfunctioning. I did abandon the layer, but created another without any heating or cooling grid.

This kind of solved the problem, However, apart from still having exceptions and black screens when working on this tile's surface grid, I have a more bothersome problem.

Each time I load the game, my main air-con network is stuck at 0/0, with an the very same exception triggered. Once I move something in the grid, it goes again. Also, I looked at the save and the HygienePipeMapComp are present for all maps but empty.

`<li Class="DubsCentralHeating.HygienePipeMapComp" />`

Is there any way to solve the on-load problem ? Once and for all. With save editing maybe. Because every so often I forget to move something and have to reload and roll back when I figure.
